lib/Dialect/Triton/Transforms/Combine.cpp

```diff
@@ -187,6 +187,29 @@ class CombineBroadcastMulReducePattern : public RewritePattern {
   }
 };
 
+// When reducing a 1D tensor the order of elements of the tensor doesn't matter.
+// Therefore we can relax the reshape to allow it to re-order elements.
+class CombineReshapeReducePatterns : public mlir::OpRewritePattern<ReshapeOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(triton::ReshapeOp reshapeOp,
+                  mlir::PatternRewriter &rewriter) const override {
+    if (reshapeOp.getAllowReorder())
+      return failure();
+    if (reshapeOp.getType().getRank() != 1)
+      return failure();
+    for (Operation *user : reshapeOp->getUsers()) {
+      if (!isa<triton::ReduceOp>(user))
+        return failure();
+    }
+    rewriter.modifyOpInPlace(reshapeOp,
+                             [&]() { reshapeOp.setAllowReorder(true); });
```

Comment:
SGTM, but is there any harm in creating two reshape ops if there are non-reduce users? I think with linear layouts an allow-reorder reshape should always be a no-op.

Reply:
Oh and this pattern works for `HistogramOp` as well.

Reply:
> SGTM, but is there any harm in creating two reshape ops if there are non-reduce users? I think with linear layouts an allow-reorder reshape should always be a no-op.

The potential harm is extra pressure if this makes the tensor exist with different layouts. We kind of already have this problem when doing ramterialization in layout propagation pass but to be conservative I would prefer avoiding it until it is proven to help.

> Oh and this pattern works for HistogramOp as well.

good point.

Reply:
> The potential harm is extra pressure if this makes the tensor exist with different layouts.

Hrm okay. It's a bit annoying that we are just papering over a flaw in the layout propagation pass, but hopefully we can resolve that at some point.

Reply:
well if we create two reshapes with and without allow_reorder then there is likely to have extra pressure whatever we do in layout propagation. What I meant is that layout propagation also does some of it but it is not great

Reply:
I might be over cautious but we can re-visit based on findings